src/modules/gfx_autoterminal.c

```diff
@@ -0,0 +1,523 @@
+                escape_code = 0;
+            }
+        }else{
+            if(c == 0x1B){
+                escape_code = 1;
+            }else{
+                tmpbuffer[0] = c;
+                tmpbuffer[1] = 0;
+                offset = write_buffer(tmpbuffer, &current_row, &current_column);
+            }
+        }
+    }
+    fclose(cout);
+    free(tmpbuffer);
+    active_shell = 0;
+    return 0;
+}
+
+static void clear_buffer() {
+    int i;
+    for (i = 0; i < max_row * max_column; ++i) {
+        buffer[i].c = ' ';
+        buffer[i].fg = fg;
+        buffer[i].bg = bg;
+    }
+}
+
+int init(int modno, char *argstr) {
+    moduleno = modno;
+
+    char *from_int = malloc(10 * sizeof(char));
+    max_row = matrix_gety() / 6;
+    max_column = matrix_getx() / 4;
+    setenv("TERMINFO", "./terminfo/", 1);
+    setenv("TERM", "autoterminal", 1);
+    snprintf(from_int, 10, "%d", max_row);
+    setenv("LINES", from_int, 1);
+    snprintf(from_int, 10, "%d", max_column);
+    setenv("COLUMNS", from_int, 1);
+    max_index = 1;
+
+    // read script
+    FILE *file;
+    char ch;
+
+    file = fopen("scripts/autoterminal.sh", "r");
+    if (file) {
+        // last line endns with eof and not new line
+        for (ch = getc(file); ch != EOF; ch = getc(file))
+            if (ch == '\n')
+                max_index++;
+        type_buffer = malloc(max_index * sizeof(char *));
+        for (type_index = 0; type_index < max_index; type_index++) {
+            type_buffer[type_index] = malloc(max_column * 3 * sizeof(char));
+        }
+        rewind(file);
+        for (type_index = 0;
+             type_index < max_index &&
+             fgets(type_buffer[type_index], max_column * 3, file) != NULL;
+             type_index++) {
+            // skip comments
+            if (type_buffer[type_index][0] == '#') {
+                type_index--;
+            }
+        };
+        fclose(file);
+    }
+
+    type_pos = 0;
+    max_index = type_index - 1;
+    type_index = 0;
+
+    buffer = malloc(max_row * max_column * sizeof(struct font_char));
+    clear_buffer();
+    if (max_index >= 0) {
+        write_buffer("$ ", &current_row, &current_column);
+    }
+
+    // 2 rows
+    if (matrix_getx() < font_height * 2)
+        return 1; // not enough X to be usable
+    // 8 characters
+    if (matrix_gety() < font_width * 8)
+        return 1; // not enough Y to be usable
+    return 0;
+}
+
+void reset(int _modno) {
+    type_index = 0;
+    type_pos = 0;
+    current_row = 0;
+    current_column = 0;
+    fg = fg_default;
+    bg = bg_default;
+    clear_buffer();
+    if (max_index >= 0) {
+        write_buffer("$ ", &current_row, &current_column);
+    }
+    nexttick = udate();
+}
+
+int draw(int _modno, int argc, char *argv[]) {
+    int x = 0;
+    int y = 0;
+    int row = 0;
+    int column = 0;
+    int pos = 0;
+    char ch[1];
+    pthread_t thread_id;
+    if (active_shell == 0) {
```

Comment:
You'll want to use a mutex (`oscore_mutex_*`, see `src/oscore.h`) to synchronize.

Reply:
I know, I was too lazy, I will look into oscore.
The text buffer also needs to be secured with mutexes.

Reply:
I would need try_lock and I don’t really see an issue with active_shell.
It does draw loops until the thread sets it to 0.

Reply:
I actually meant events (`oscore_event_*`) here, but upon reflection there is no loop here that the compiler can see, so it has no reason to optimize the read out.
The reason I wrote that was because it's good practice to not use simple variables to syncronize as a compiler might optimize it out in loops if it is invoked with higher optimization levels, for example.